Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -1429,6 +1468,114 @@ TEST_F(SerializerTest, EncodesResumeTokens) {
+  proto.mutable_target_change()->add_target_ids(1);
+  proto.mutable_target_change()->add_target_ids(2);
+  proto.mutable_target_change()->set_resume_token("resume_token");
+
+  SCOPED_TRACE("DecodesListenResponseWithTargetChange");
+  ExpectWatchChangeDeserializationRoundTrip(model, proto);
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithDocumentChange) {
+  SnapshotVersion version = Version(123456789L);
+  DocumentWatchChange model(
+      {1, 3}, {2, 4}, Key("one/two/three/four"),
+      Doc("one/two/three/four", 123456789L, Map("foo", "bar")));
+
+  v1::ListenResponse proto;
+
+  auto document_change = proto.mutable_document_change();
+  document_change->mutable_document()->set_name(
+      "projects/p/databases/d/documents/one/two/three/four");
+  document_change->mutable_document()->mutable_update_time()->set_seconds(
+      version.timestamp().seconds());
+  document_change->mutable_document()->mutable_update_time()->set_nanos(
+      version.timestamp().nanoseconds());
+  (*document_change->mutable_document()->mutable_fields())["foo"] =
+      ValueProto("bar");
+
+  document_change->add_target_ids(1);
+  document_change->add_target_ids(3);
+  document_change->add_removed_target_ids(2);
+  document_change->add_removed_target_ids(4);
+
+  SCOPED_TRACE("DecodesListenResponseWithDocumentChange");
+  ExpectWatchChangeDeserializationRoundTrip(model, proto);
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithDocumentDelete) {
+  DocumentWatchChange model({}, {1}, Key("one/two/three/four"),
+                            DeletedDoc("one/two/three/four"));
+
+  v1::ListenResponse proto;
+
+  auto document_delete = proto.mutable_document_delete();
+  document_delete->set_document(
+      "projects/p/databases/d/documents/one/two/three/four");
+
+  document_delete->add_removed_target_ids(1);
+
+  SCOPED_TRACE("DecodesListenResponseWithDocumentDelete");
+  ExpectWatchChangeDeserializationRoundTrip(model, proto);
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithDocumentRemove) {
+  DocumentWatchChange model({}, {1, 2}, Key("one/two/three/four"),
+                            absl::nullopt);
+
+  v1::ListenResponse proto;
+
+  auto document_remove = proto.mutable_document_remove();
+  document_remove->set_document(
+      "projects/p/databases/d/documents/one/two/three/four");
+
+  document_remove->add_removed_target_ids(1);
+  document_remove->add_removed_target_ids(2);
+
+  SCOPED_TRACE("DecodesListenResponseWithDocumentRemove");
+  ExpectWatchChangeDeserializationRoundTrip(model, proto);
+}
+
+TEST_F(SerializerTest, DecodesListenResponseWithExistenceFilter) {
+  ExistenceFilterWatchChange model(ExistenceFilter(2), 100);
+
+  v1::ListenResponse proto;
+
+  proto.mutable_filter()->set_count(2);
+  proto.mutable_filter()->set_target_id(100);
+
+  SCOPED_TRACE("DecodesListenResponseWithExistenceFilter");
+  ExpectWatchChangeDeserializationRoundTrip(model, proto);
+}
+
```

Comment:
Can you also port [tests](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Example/Tests/Remote/FSTSerializerBetaTests.mm#L490-L513) that exercise `DecodeMutationResult`?